orttraining/orttraining/python/training/pytorch_trainer_options.py

```diff
@@ -0,0 +1,376 @@
+            Users can also instantiate :py:class:`.DynamicLossScaler` and
+            override its parameters. Lastly, a completely new implementation
+            can be specified by extending :py:class:`.LossScaler` class from scratch
+        utils (dict):
+            miscellaneous options
+        utils.grad_norm_clip (bool, default is False):
+            enables gradient norm clipping for 'AdamOptimizer' and 'LambOptimizer'
+        _internal_use (dict):
+            internal, possibly undocumented, options that might be removed in the next release
+        _internal_use.frozen_weights (list, []):
+            list of model parameters to freeze (stop training)
+        _internal_use.enable_internal_postprocess (bool, default is True):
+            enable internal internal post processing of the ONNX model
+        _internal_use.extra_postprocess (callable, default is None)
+            a functor to postprocess the ONNX model.
+            It does not override :py:attr:`._internal_use.enable_internal_postprocess`, but complement it
+
+    Example:
+        .. code-block:: python
+
+            opts = PytorchTrainerOptions({
+                               'batch' : {
+                                   'gradient_accumulation_steps' : 128
+                               },
+                               'device' : {
+                                   'id' : 'cuda:0',
+                                   'mem_limit' : 2*1024*1024*1024,
+                               },
+                               'lr_scheduler' : optim.lr_scheduler.LinearWarmupLRScheduler(),
+                               'mixed_precision' : {
+                                   'enabled': True,
+                                   'loss_scaler': amp.LossScaler(loss_scale=float(1 << 16))
+                               }
+            })
+            fp16_enabled = opts.mixed_precision.enabled
+     """
+
+    def __init__(self, options, _validate=True):
+        # Keep a copy of original input for debug
+        self._original_opts = dict(options)
+
+        # Add an empty dictionary for non specified nested dicts
+        subgroups = [k for k, v in _PYTORCH_TRAINER_OPTIONS_SCHEMA.items()
+                     if isinstance(v, dict) and 'type' in v and v['type'] == 'dict']
+        self._validated_opts = dict(self._original_opts)
+        if _validate:
+            for name in subgroups:
+                if name not in self._validated_opts:
+                    self._validated_opts[name] = {}
+
+            # Validates user input
+            validator = PytorchTrainerOptionsValidator(_PYTORCH_TRAINER_OPTIONS_SCHEMA)
+            self._validated_opts = validator.validated(self._validated_opts)
+            if self._validated_opts is None:
+                self._validated_opts = validator.errors
+            _validate = False
+
+        # Convert dict in object
+        for k, v in self._validated_opts.items():
+            setattr(self, k, self._wrap(v, _validate))
```

Comment:
is `_validate` here always False?

Reply:
That was one work-around that I found to mimic static class variables :( basically _validate=True when `PytorchTrainerOptions` is constructed. All recursive calls to its sub dictionaries must have _validate=False.

An alternate design is to create two classes, `PytorchTrainerOptions` and `_PytorchTrainerOptionsInternal`. In that approach, The former would call validate where the second wouldn't. I started with the current approach to minimize the coding, but if you feel strong enough, w can change this.

Reply:
Refactored the code into 2 classes to fix this